OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/OpenPgpServiceKeyIdExtractor.java

```diff
@@ -53,150 +47,223 @@ private OpenPgpServiceKeyIdExtractor(ContentResolver contentResolver,
     }
 
 
-    KeyIdResult returnKeyIdsFromIntent(Intent data, boolean askIfNoUserIdsProvided) {
-        HashSet<Long> encryptKeyIds = new HashSet<>();
-
+    KeyIdResult returnKeyIdsFromIntent(Intent data, boolean askIfNoUserIdsProvided, String callingPackageName) {
         boolean hasKeysFromSelectPubkeyActivity = data.hasExtra(OpenPgpApi.EXTRA_KEY_IDS_SELECTED);
+
+        KeyIdResult result;
         if (hasKeysFromSelectPubkeyActivity) {
+            HashSet<Long> encryptKeyIds = new HashSet<>();
             for (long keyId : data.getLongArrayExtra(OpenPgpApi.EXTRA_KEY_IDS_SELECTED)) {
                 encryptKeyIds.add(keyId);
             }
+            result = createKeysOkResult(encryptKeyIds, false);
         } else if (data.hasExtra(OpenPgpApi.EXTRA_USER_IDS) || askIfNoUserIdsProvided) {
             String[] userIds = data.getStringArrayExtra(OpenPgpApi.EXTRA_USER_IDS);
-            boolean isOpportunistic = data.getBooleanExtra(OpenPgpApi.EXTRA_OPPORTUNISTIC_ENCRYPTION, false);
-            KeyIdResult result = returnKeyIdsFromEmails(data, userIds, isOpportunistic);
-
-            if (result.mResultIntent != null) {
-                return result;
-            }
-            encryptKeyIds.addAll(result.mKeyIds);
+            result = returnKeyIdsFromEmails(data, userIds, callingPackageName);
+        } else {
+            result = createNoKeysResult();
         }
 
         // add key ids from non-ambiguous key id extra
         if (data.hasExtra(OpenPgpApi.EXTRA_KEY_IDS)) {
+            HashSet<Long> explicitKeyIds = new HashSet<>();
             for (long keyId : data.getLongArrayExtra(OpenPgpApi.EXTRA_KEY_IDS)) {
-                encryptKeyIds.add(keyId);
+                explicitKeyIds.add(keyId);
             }
+            result = result.withExplicitKeyIds(explicitKeyIds);
         }
 
-        if (encryptKeyIds.isEmpty()) {
-            Intent result = new Intent();
-            result.putExtra(OpenPgpApi.RESULT_ERROR,
-                    new OpenPgpError(OpenPgpError.NO_USER_IDS, "No encryption keys or user ids specified!" +
-                            "(pass empty user id array to get dialog without preselection)"));
-            result.putExtra(OpenPgpApi.RESULT_CODE, OpenPgpApi.RESULT_CODE_ERROR);
-            return new KeyIdResult(result);
-        }
-
-        return new KeyIdResult(encryptKeyIds);
+        return result;
     }
 
-    private KeyIdResult returnKeyIdsFromEmails(Intent data, String[] encryptionUserIds, boolean isOpportunistic) {
+    private KeyIdResult returnKeyIdsFromEmails(Intent data, String[] encryptionUserIds, String callingPackageName) {
```

Comment:
This method no longer determines the recipient status itself, but instead calls out to KeychainExternalProvider. The "callingPackageName" parameter allows making these queries calling package sensitive, which is not used right now but can be for application-specific key/identity bindings at a later point.